### ISSUE TITLE
Dataset read error with 3D vector

### ISSUE THREAD

**Issue by Mrporkk** [BlueBrain/HighFive#96]
When I compile this code I get error with this,but if I notes "dataset1.read(data);" this code,it compiles successful, and show label.size() is right. How to solve this problem?

# SOURCE CODE
std::vector<std::vector<std::vector<float>>> data;
std::vector<int> label;
std::string filename = "/path/to/data/h5_data/110020171219_h5/0.h5";
File file(filename, File::ReadOnly);
DataSet dataset1 = file.getDataSet("data");
dataset1.read(data);
DataSet dataset2 = file.getDataSet("label");
dataset2.read(label);
std::cout << label.size() << std::endl;

# ERROR
In file included from /path/to/CLionProjects/Create_hdf5/src/highfive/bits/../bits/H5Attribute_misc.hpp:29:0,
                 from /path/to/CLionProjects/Create_hdf5/src/highfive/bits/../H5Attribute.hpp:73,
                 from /path/to/CLionProjects/Create_hdf5/src/highfive/bits/H5Annotate_traits_misc.hpp:18,
                 from /path/to/CLionProjects/Create_hdf5/src/highfive/bits/H5Annotate_traits.hpp:74,
                 from /path/to/CLionProjects/Create_hdf5/src/highfive/H5File.hpp:17,
                 from/path/to/CLionProjects/Create_hdf5/src/main.cpp:7:
/path/to/CLionProjects/Create_hdf5/src/highfive/bits/../bits/H5Converter_misc.hpp: In instantiation of ‘typename std::vector<_RealType>::iterator HighFive::details::single_buffer_to_vectors(typename std::vector<_RealType>::iterator, typename std::vector<_RealType>::iterator, const std::vector<long unsigned int>&, std::size_t, std::vector<WeightType>&) [with T = float; U = std::vector<std::vector<float> >; typename std::vector<_RealType>::iterator = __gnu_cxx::__normal_iterator<float*, std::vector<float> >; std::size_t = long unsigned int]’:
/path/to/CLionProjects/Create_hdf5/src/highfive/bits/../bits/H5Converter_misc.hpp:272:64:   required from ‘void HighFive::details::data_converter<std::vector<_RealType>, typename HighFive::details::enable_if<HighFive::details::is_container< <template-parameter-1-1> >::value>::type>::process_result(std::vector<_RealType>&) [with T = std::vector<std::vector<float> >; typename HighFive::details::enable_if<HighFive::details::is_container< <template-parameter-1-1> >::value>::type = void]’
/path/to/CLionProjects/Create_hdf5/src/highfive/bits/../bits/H5Slice_traits_misc.hpp:184:5:   required from ‘void HighFive::SliceTraits<Derivate>::read(T&) const [with T = std::vector<std::vector<std::vector<float> > >; Derivate = HighFive::Selection]’
/path/to/CLionProjects/Create_hdf5/src/main.cpp:113:61:   required from here
/path/to/CLionProjects/Create_hdf5/src/highfive/bits/../bits/H5Converter_misc.hpp:105:69: error: no matching function for call to ‘single_buffer_to_vectors(std::vector<float>::iterator&, std::vector<float>::iterator&, const std::vector<long unsigned int>&, std::size_t, std::vector<std::vector<float> >&)’
                                                 current_dim + 1, *it);
                                                                     ^
/path/to/CLionProjects/Create_hdf5/src/highfive/bits/../bits/H5Converter_misc.hpp:105:69: note: candidates are:
/path/to/CLionProjects/Create_hdf5/src/highfive/bits/../bits/H5Converter_misc.hpp:81:1: note: typename std::vector<_RealType>::iterator HighFive::details::single_buffer_to_vectors(typename std::vector<_RealType>::iterator, typename std::vector<_RealType>::iterator, const std::vector<long unsigned int>&, std::size_t, std::vector<_RealType>&) [with T = std::vector<float>; typename std::vector<_RealType>::iterator = __gnu_cxx::__normal_iterator<std::vector<float>*, std::vector<std::vector<float> > >; std::size_t = long unsigned int]
 single_buffer_to_vectors(typename std::vector<T>::iterator begin_buffer,
 ^
/path/to/CLionProjects/Create_hdf5/src/highfive/bits/../bits/H5Converter_misc.hpp:81:1: note:   no known conversion for argument 1 from ‘std::vector<float>::iterator {aka __gnu_cxx::__normal_iterator<float*, std::vector<float> >}’ to ‘std::vector<std::vector<float> >::iterator {aka __gnu_cxx::__normal_iterator<std::vector<float>*, std::vector<std::vector<float> > >}’
/path/to/CLionProjects/Create_hdf5/src/highfive/bits/../bits/H5Converter_misc.hpp:94:1: note: template<class T, class U> typename std::vector<_RealType>::iterator HighFive::details::single_buffer_to_vectors(typename std::vector<_RealType>::iterator, typename std::vector<_RealType>::iterator, const std::vector<long unsigned int>&, std::size_t, std::vector<WeightType>&)
 single_buffer_to_vectors(typename std::vector<T>::iterator begin_buffer,
 ^
/path/to/CLionProjects/Create_hdf5/src/highfive/bits/../bits/H5Converter_misc.hpp:94:1: note:   template argument deduction/substitution failed:
/path/to/CLionProjects/Create_hdf5/src/highfive/bits/../bits/H5Converter_misc.hpp:105:69: note:   couldn't deduce template parameter ‘T’
                                                 current_dim + 1, *it);
                                                                     ^

**Comment by jrs65** [comment]
I'm pretty sure this is because you never actually give a type to either of the templated `std::vector`s. (i.e. line 1 and 2). Usually it would be something like `std::vector<std::vector<std::vector<float>>> data;`

**Comment by Mrporkk** [comment]
Thank you jrs65,
Is my mistake to copy the error code , this is the original code.
 
std::vector<std::vector<std::vector<float>>> data;
std::vector<int> label;

**Comment by Mrporkk** [comment]
Thank you jrs,
Is my mistake to copy the error code because of I don't click the "insert code" button,my original code is this.
```
std::vector<std::vector<std::vector<float>>> data;
std::vector<int> label;
```

**Comment by srijitpaul** [comment]
Dear all,
I am facing the same compilation problem with reading a 3-D std::complex<double> vector,
Here is the code snippet,
`      std::vector<std::vector<std::vector<std::complex<double>>>> result;
      dataset.select({gresult.second, momresult.second, 0},{0, 0, 48}).read(result);
`
The compilation error is the following,
/global/homes/s/srijitp/spaul/projects/production_pion_nucleon/qlua-scripts_all_hdf5/recombination/Gamma_testers/include/highfive/bits/../bits/H5Converter_misc.hpp(119): error: no instance of overloaded function "HighFive::details::single_buffer_to_vectors" matches the argument list
            argument types are: (__gnu_cxx::__normal_iterator<std::complex<double> *, std::vector<std::complex<double>, std::allocator<std::complex<double>>>>, __gnu_cxx::__normal_iterator<std::complex<double> *, std::vector<std::complex<double>, std::allocator<std::complex<double>>>>, const std::vector<size_t, std::allocator<size_t>>, unsigned long, std::vector<std::vector<std::complex<double>, std::allocator<std::complex<double>>>, std::allocator<std::vector<std::complex<double>,
                      std::allocator<std::complex<double>>>>>)
          begin_buffer = single_buffer_to_vectors(begin_buffer, end_buffer, dims,
                         ^
/global/homes/s/srijitp/spaul/projects/production_pion_nucleon/qlua-scripts_all_hdf5/recombination/Gamma_testers/include/highfive/bits/../bits/H5Converter_misc.hpp(108): note: this candidate was rejected because at least one template argument could not be deduced
  single_buffer_to_vectors(typename std::vector<T>::iterator begin_buffer,
  ^
/global/homes/s/srijitp/spaul/projects/production_pion_nucleon/qlua-scripts_all_hdf5/recombination/Gamma_testers/include/highfive/bits/../bits/H5Converter_misc.hpp(94): note: this candidate was rejected because arguments do not match
  single_buffer_to_vectors(typename std::vector<T>::iterator begin_buffer,
  ^
          detected during:
            instantiation of "std::vector<T, std::allocator<T>>::iterator HighFive::details::single_buffer_to_vectors<T,U>(std::vector<T, std::allocator<T>>::iterator, std::vector<T, std::allocator<T>>::iterator, const std::vector<size_t={unsigned long}, std::allocator<size_t={unsigned long}>> &, size_t={unsigned long}, std::vector<U, std::allocator<U>> &) [with T=std::complex<double>, U=std::vector<std::vector<std::complex<double>, std::allocator<std::complex<double>>>,
                      std::allocator<std::vector<std::complex<double>, std::allocator<std::complex<double>>>>>]" at line 301
            instantiation of "void HighFive::details::data_converter<std::vector<T, std::allocator<T>>, std::enable_if<HighFive::details::is_container<T>::value, void>::type>::process_result(std::vector<T, std::allocator<T>> &) [with T=std::vector<std::vector<std::complex<double>, std::allocator<std::complex<double>>>, std::allocator<std::vector<std::complex<double>, std::allocator<std::complex<double>>>>>]" at line 188 of
                      "/global/homes/s/srijitp/spaul/projects/production_pion_nucleon/qlua-scripts_all_hdf5/recombination/Gamma_testers/include/highfive/bits/../bits/H5Slice_traits_misc.hpp"
            instantiation of "void HighFive::SliceTraits<Derivate>::read(T &) const [with Derivate=HighFive::Selection, T=std::vector<std::vector<std::vector<std::complex<double>, std::allocator<std::complex<double>>>, std::allocator<std::vector<std::complex<double>, std::allocator<std::complex<double>>>>>, std::allocator<std::vector<std::vector<std::complex<double>, std::allocator<std::complex<double>>>, std::allocator<std::vector<std::complex<double>,
                      std::allocator<std::complex<double>>>>>>>]" at line 311 of "/global/homes/s/srijitp/spaul/projects/production_pion_nucleon/qlua-scripts_all_hdf5/recombination/Gamma_testers/src/main10.cpp"

compilation aborted for /global/homes/s/srijitp/spaul/projects/production_pion_nucleon/qlua-scripts_all_hdf5/recombination/Gamma_testers/src/main10.cpp (code 2)
CMakeFiles/h5read_mic-knl.dir/build.make:62: recipe for target 'CMakeFiles/h5read_mic-knl.dir/src/main10.cpp.o' failed
make[3]: *** [CMakeFiles/h5read_mic-knl.dir/src/main10.cpp.o] Error 2
CMakeFiles/Makefile2:363: recipe for target 'CMakeFiles/h5read_mic-knl.dir/all' failed
make[2]: *** [CMakeFiles/h5read_mic-knl.dir/all] Error 2
CMakeFiles/Makefile2:375: recipe for target 'CMakeFiles/h5read_mic-knl.dir/rule' failed
make[1]: *** [CMakeFiles/h5read_mic-knl.dir/rule] Error 2
Makefile:268: recipe for target 'h5read_mic-knl' failed
make: *** [h5read_mic-knl] Error 2

**Comment by ferdonline** [comment]
fixed by #191
